### PR TITLE
Use npm to manage fomantic and only build needed components

### LINKED ISSUE #9562: Custom fonts URL on fomantic
Currently, fomantic will load fonts from https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic&subset=latin&display=swap. That should be configurable.

### PR DESCRIPTION
**Pull request by lunny**
Fix #9562, #9415

- Removed semantic.min.css from git and use npm to build fomantic. 
- Only load needed semantic components. Before this PR, `semantic.min.css` is 1.29MB, after ~650KB.
- Don't load fonts from google sites since it's slow from some location visit. And the lato fonts has been loaded by gitea itself.

### CODE REVIEW

**Review by bagasme — COMMENTED** [review]
Huge PR, right?

**Comment by lunny** [comment]
@bagasme I don't think so. Most changes are to delete semantic-ui vendor files and add semantic-ui sources.

**Review by lafriks — CHANGES_REQUESTED** [review]

File: package.json
Comment (lunny) [reply package.json]: @lafriks I just follow https://fomantic-ui.com/introduction/build-tools.html to use gulp. If we want to use webpack, I think we should do that another PRs.

**Comment by lunny** [comment]
@lafriks Done. I also moved `fomantic-ui` to `devDependencies`.

**Comment by 6543** [comment]
@lunny is there a optio to only build fomatic if something has changed?

it is build every time you build gitea witch is anoing

**Comment by lafriks** [comment]
Fomatic-ui should be dependency, and gulp devDependency

File: package.json
Comment (silverwind) [inline package.json]: Maybe move it to `dependencies` given that it is code actually used on the frontend.

File: package.json
Comment (silverwind) [inline package.json]: We can probably do without gulp as a direct dependency as fomantic-ui already depends on it, guaranteeing it to be there.

File: Makefile
Comment (silverwind) [inline Makefile:490]: Maybe it is possible to have this as a non-phony target that depends on `package-lock.json`, so it would only run if dependencies change.

File: package.json
Comment (silverwind) [reply package.json]: Gulp build is necessary to build the fomantic bundle I'm afraid, but we can tweak the Makefile so it only runs when necessary. Optionally, the resulting files could then be integrated in the webpack bundle as most pages do need it (thought, not all).

File: Makefile
Comment (silverwind) [reply Makefile:490]: Also, I'd suggest the target to be adjusted to provide one or more output files, like `public/fomantic/semantic.min.js` and `public/fomantic/semantic.min.css`. We can still have a phony `fomantic` target that depends on that file to trigger it on-demand.

File: semantic.json
Comment (silverwind) [inline semantic.json]: trailing slash for consistency?

File: semantic.json
Comment (silverwind) [inline semantic.json:58]: sort alphabetically

**Review by das7pad — COMMENTED** [review]
@lunny @6543 I managed to devendor fomantic (8MB on disk) and implement the lazy build.

Feel free to cherry pick https://github.com/das7pad/gitea/commit/6a012e9babb3b7b7ca7e94120978b2be812aceeb

**Comment by lunny** [comment]
@das7pad Thanks! @lafriks @silverwind done.

**Comment by silverwind** [comment]
Please remove the `gulp` dependency.

Not sure I like the approach of writing into `node_modules`, I think it may be better to do it outside and treat it as read-only.

**Comment by das7pad** [comment]
> Please remove the `gulp` dependency.

Can you help me understand why it should be removed?

> Not sure I like the approach of writing into `node_modules`, I think it may be better to do it outside and treat it as read-only.

I do not like it either, but their imports dictate that this config file is placed in the source code:
https://github.com/fomantic/Fomantic-UI/blob/037a0402362fa2ffa44e434fe7e362c8cb753bb8/src/definitions/globals/reset.less#L18

As noted in the TODO, we can work around that with an alias/mock, but I could not find an easy solution for that using the gulp build tool. We could fix the fomantic source too, but i'm not sure it is worth the effort.

File: Makefile
Comment (das7pad) [inline Makefile]: @lunny Missing files are not subject to diffs and produce 0 output lines on stdout, hence the broken build.

There is a `-N` option that treats missing files as empty.
We can work with the return code too and the target body can be simplified to `diff -qN $< $@ || cp -f $< $@` (`-q` suppresses the diff).

<details>
<summary> full diff </summary>
  
```diff
diff --git a/Makefile b/Makefile
index 56152c53f..7b2ce3e34 100644
--- a/Makefile
+++ b/Makefile
@@ -489,10 +489,7 @@ node_modules/fomantic-ui/src: node-check node_modules
 # TODO(das7pad): replace with resolve.alias when building fomantic with webpack
 node_modules/fomantic-ui/src/theme.config: node_modules/fomantic-ui/src
 node_modules/fomantic-ui/src/theme.config: web_src/fomantic/theme.config.less
-       @diff=$$(diff '$<' '$@'); \
-       if [ -n "$$diff" ]; then \
-               cp -f $< $@; \
-       fi;
+       diff -qN $< $@ || cp -f $< $@
 
 $(FOMANTIC_DEST_DIR): node_modules/fomantic-ui/src
 $(FOMANTIC_DEST_DIR): node_modules/fomantic-ui/src/theme.config
```
</details>

Comment (lunny) [reply Makefile]: @das7pad I have changed that to 
```makefile
ifeq (,$(wildcard $@))
	cp $< $@
endif
```

**Comment by silverwind** [comment]
> Can you help me understand why it should be removed?

Because it already comes in [as a dependency of fomantic](https://github.com/fomantic/Fomantic-UI/blob/037a0402362fa2ffa44e434fe7e362c8cb753bb8/package.json#L43). If we depend on a specific version of it directly and fomantic updates their version, we'd end up with two versions which is not ideal (not sure which one would end up in the `.bin` scripts in that case).

File: Makefile
Comment (das7pad) [reply Makefile]: Now it will not be updated upon changing the local file in `web_src`.

Let's step back and just use the timestamp based logic of make to determine the need for a new copy call. Just call `cp -f $< $@` in the target body.

**Comment by das7pad** [comment]
Thanks @silverwind !
 
> > Can you help me understand why it should be removed?
> 
> Because it already comes in [as a dependency of fomantic](https://github.com/fomantic/Fomantic-UI/blob/037a0402362fa2ffa44e434fe7e362c8cb753bb8/package.json#L43). If we depend on a specific version of it directly and fomantic updates their version, we'd end up with two versions which is not ideal (not sure which one would end up in the `.bin` scripts in that case).
> 
> If you really want to keep it, I'd suggest depending on version `*` so the more specific definition in fomantic would take precedence.

Our version would go into `.bin`, theirs in their `node_modules`.
I do not think that having two versions is a problem per se. gulp4s' cli could be perfectly compatible with the gulp5 modules that their build-file loads. But gulp5 could also drop the `-f` option or expect a sub command to trigger the full build process. I would say the it is up to the developer who proposes a new fomantic (or gulp) version to make sure it is compatible with the build tooling.

I agree that we may be better off with a more loose version. `^4.0.0` would leave full room for minor/patch version upgrades but retain the major version at 4. IIRC `*` does not play nice with lock files and we do not want to end up with gulp9 when fomantic is stuck at 4.

**Comment by silverwind** [comment]
In the end, it doesn't really matter much because we already lock dependencies. Still, I would recommend removing it from our dependencies, the [fomantic docs](https://fomantic-ui.com/introduction/getting-started.html#installing-via-npm) also don't recommend adding gulp to the project dependencies.

**Comment by das7pad** [comment]
Going by the docs, we would vendor fomatic too ;)

<details>
<summary> $ npm install fomantic-ui </summary>

```bash
$ cd `mktemp -d`
$ nvm use 12.14.1
Now using node v12.14.1 (npm v6.13.4)
$ npm init
---
SNIP
---
$ npm install fomantic-ui
npm WARN deprecated gulp-util@3.0.8: gulp-util is deprecated - replace it, following the guidelines at https://medium.com/gulpjs/gulp-util-ca3b1f9f9ac5

> fomantic-ui@2.8.3 install /tmp/tmp.lLcsEhjjXz/node_modules/fomantic-ui
> gulp install

[14:06:57] Using gulpfile /tmp/tmp.lLcsEhjjXz/node_modules/fomantic-ui/gulpfile.js
[14:06:57] Starting 'install'...

[14:06:57] Starting 'run setup'...
? Set-up Semantic UI Automatic (Use default locations and all components)
? We detected you are using NPM Nice! Is this your project folder? /tmp/tmp.lLcsEhjjXz Yes
? Where should we put Semantic UI inside your project? semantic/
[14:07:43] Finished 'run setup' after 46 s
[14:07:43] Starting 'create install files'...

Installing
------------------------------
Installing to semantic/
Copying UI definitions
Copying UI themes
Copying gulp tasks
Adding theme files
Creating gulpfile.js
Creating site theme folder /tmp/tmp.lLcsEhjjXz/semantic/src/site/
[14:07:43] Starting 'create theme.config'...
Adjusting @siteFolder to:  site/
Creating src/theme.config (LESS config) /tmp/tmp.lLcsEhjjXz/semantic/src/theme.config
[14:07:43] Finished 'create theme.config' after 25 ms
[14:07:43] Starting 'create semantic.json'...
Creating config file (semantic.json) /tmp/tmp.lLcsEhjjXz/semantic.json
[14:07:43] Finished 'create semantic.json' after 18 ms
[14:07:43] Finished 'create install files' after 148 ms
[14:07:43] Starting 'clean up install'...

 Setup Complete! 
 Installing Peer Dependencies. Please refrain from ctrl + c... 
 After completion navigate to semantic/ and run "gulp build" to build
[14:07:43] Finished 'clean up install' after 680 μs
[14:07:43] Finished 'install' after 46 s
npm notice created a lockfile as package-lock.json. You should commit this file.
---
SNIP
---

+ fomantic-ui@2.8.3
added 750 packages from 460 contributors and audited 14412 packages in 109.696s

8 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

$ tree -L 2
.
├── node_modules
│   ├── abbrev
---
SNIP
---
│   └── yargs-parser
├── package.json
├── package-lock.json
├── semantic
│   ├── gulpfile.js
│   ├── src
│   └── tasks
└── semantic.json

556 directories, 4 files
```
</details>

**Comment by silverwind** [comment]
I think you're missing the `progress` component. I see a ` TypeError: $(...).progress is not a function` error when loading the frontpage. Also, the right side repo/org box appears missing on the frontpage, not sure if same cause.

File: Makefile
Comment (silverwind) [inline Makefile]: Not sure I understand what's happening here or why this `ifeq` is not indented but wouldn't something like this be clearer?

```` make
node_modules/fomantic-ui/src/theme.config: node_modules web_src/fomantic/theme.config.less
  cp web_src/fomantic/theme.config.less $@

$(FOMANTIC_DEST_DIR): node_modules/fomantic-ui/src/theme.config
	npx gulp -f node_modules/fomantic-ui/gulpfile.js build
````

edit: see suggestion below

**Comment by silverwind** [comment]
Suggestion for makefile simplification:

```` make
.PHONY: fomantic
fomantic: node-check node_modules $(FOMANTIC_DEST_DIR)

$(FOMANTIC_DEST_DIR): semantic.json web_src/fomantic/theme.config.less
  cp web_src/fomantic/theme.config.less node_modules/fomantic-ui/src/theme.config
  npx gulp -f node_modules/fomantic-ui/gulpfile.js build
````

Should be a bit clearer that way and it add the important ``semantic.json` dependency. I see no reason to keep the `node_modules/fomantic-ui/src` target because it should be covered by `node_modules`.

File: Makefile
Comment (das7pad) [reply Makefile]: ~`node_modules` is a phony target, hence it always runs and any following parents will always run too -- the fomantic assets are always rebuilt. We need an intermediate target that is not phony but backed by a file/dir on disk.~ Edit: `node-check` is the phony target, not `node_modules`.

There is a previous thread regarding the target body, but it is marked as outdated by Github: https://github.com/go-gitea/gitea/pull/9561#discussion_r363738468

Initially I symlinked the two - @lunny why was this changed in the first place?
>  why this `ifeq` is not indented

This will render the `...theme.config` target as empty using `Makefile` logic and skip the `cp` call.

Comment (silverwind) [reply Makefile]: `cp` seems fine for platforms/file systems that do not support symlinks. It's more compatible that way.

File: package.json
Comment (silverwind) [inline package.json]: Still need to move this to `dependencies`.

File: Makefile
Comment (silverwind) [inline Makefile]: useless whitespace change

File: package.json
Comment (das7pad) [reply package.json]: Now I see where this is going: `depedencies` for building `gitea` and the `devDependencies` for ci (eslint/stylelint).

`webpack` and all the plugins/loader should be moved into `dependencies` then - subject to another PR of course.

Comment (silverwind) [reply package.json]: Yes, ideally we'd use `npm install --only=prod` for the build and `npm install --only=dev` for tests but that will require a few more changes in the way tests run, yeah.

Right now I'd like to reserve `dependencies` to JS loaded at runtime in the browser.

File: Makefile
Comment (lunny) [reply Makefile]: @silverwind this will copy the file only the source file changed(they are different).

**Comment by Jookia** [comment]
My computer's too slow to load the diff on github and I'm about to sleep, so apologies for the dumb question: Where is fomantic pulled in from? Is it being vendored in the source tree as part of the gitea repo, or is it being downloaded at build time? If it's being downloaded at build time, could I request gitea fork fomantic and use that for now so I can shove fomantic UI changes as PRs there and gitea can use them without having to wait an entire release cycle?

**Comment by lunny** [comment]
@Jookia Fomantic will be downloaded via npm when build gitea. I prefer not to change it directly for easy upgrade in future. How about send PRs to upstream project?

**Comment by Jookia** [comment]
On Fri, Jan 10, 2020 at 07:29:52AM -0800, Lunny Xiao wrote:
> @Jookia Fomantic will be downloaded via npm when build gitea. I prefer not to change it directly for easy upgrade in future. How about send PRs to upstream project?

Sure. But what if changes to make Gitea accessible effectively get ping
ponged between Fomantic and Gitea?

**Comment by zeripath** [comment]
We shouldn't and can't just throw away @Jookia's work on the drop-down, accessibility is for everyone and is a legal requirement for using Gitea in several jurisdictions. 

Yup some of this needs to go upstream but I wouldn't be surprised if some of these changes were related to the way we use drop-down and so aren't necessarily upstreamable. We have a working solution and maintaining a patch, although annoying, is not really very different from some of our other more complex js. Albeit the tooling isn't really there.

@Jookia have you been able to open an upstream PR to fomantic yet?

**Comment by Jookia** [comment]
Not yet. I've been setting up my dev environment again and have gotten back to work.

File: semantic.json
Comment (silverwind) [inline semantic.json:48]:
```suggestion
    "popup",
    "progress",
    "reset",
```

Comment (silverwind) [reply semantic.json:48]: This fixes the frontpage for me.

**Comment by lunny** [comment]
@zeripath @Jookia I don't know if this PR will break `web_src/js/semanticDropdown.js`. Could you help to test that?

**Comment by lunny** [comment]
@lafriks @silverwind Please review again.

**Comment by Jookia** [comment]
On Sat, Jan 11, 2020 at 06:31:31PM -0800, Lunny Xiao wrote:
> @zeripath @Jookia I don't know if this PR will break `web_src/js/semanticDropdown.js`. Could you help to test that?

Just check that the top of site dropdown menus still have aria- attributes

File: semantic.json
Comment (silverwind) [inline semantic.json:31]:
```suggestion
    "dimmer",
    "divider",
```

Comment (silverwind) [reply semantic.json:31]: Fixes `Modal: UI Dimmer, a required component is not included in this page` seen on page user settings -> security.

**Review by silverwind — CHANGES_REQUESTED** [review]

**Comment by silverwind** [comment]
> Just check that the top of site dropdown menus still have aria- attributes

Can confirm aria attributes still present.

**Comment by Jookia** [comment]
On Sun, Jan 12, 2020 at 07:28:56AM -0800, silverwind wrote:
> > Just check that the top of site dropdown menus still have aria- attributes
> 
> Can confirm aria attributes still present.

Then it should be fine. :)

**Review by silverwind — APPROVED** [review]
LGTM once above suggestion is merged.

**Review by 6543 — APPROVED** [review]
now make take 2s longer .. this is ok

**Comment by lunny** [comment]
need @lafriks approval.

File: package.json
Comment (silverwind) [inline package.json]: Please remove again.

Comment (zeripath) [reply package.json]:
```suggestion
```

Comment (silverwind) [reply package.json]: Don't forget to regenerate `package-lock-json`, e.g. `npm i --package-lock-only`.

Comment (lunny) [reply package.json]: done.

**Comment by lafriks** [comment]
gulp should be added as devDependency

**Comment by lunny** [comment]
> In the end, it doesn't really matter much because we already lock dependencies. Still, I would recommend removing it from our dependencies, the [fomantic docs](https://fomantic-ui.com/introduction/getting-started.html#installing-via-npm) also don't recommend adding gulp to the project dependencies.

@lafriks as @silverwind mentioned. fomantic-ui has a dependency on gulp already.

**Review by lafriks — APPROVED** [review]

**Comment by silverwind** [comment]
@lafriks we discussed this a few times already. Given that `gulp` comes as a second-level dependency, I think it's better to not directly depend on it to allow fomantic-ui choose their appropriate version.

**Comment by silverwind** [comment]
@lunny: any specific reason why you added `fomantic` as prereq to `js` and `css` steps? I would prefer to be able to run webpack without fomantic building.

**Comment by lunny** [comment]
@silverwind I just think, maybe some js or css files depend on fomantic since it's a basic framework. So I did that. If that's not a thing, I think we can change that.

**Pull request merged**
